core/src/main/scala/org/apache/spark/io/CompressionCodec.scala

```diff
@@ -154,8 +154,53 @@ class SnappyCompressionCodec(conf: SparkConf) extends CompressionCodec {
 
   override def compressedOutputStream(s: OutputStream): OutputStream = {
     val blockSize = conf.getSizeAsBytes("spark.io.compression.snappy.blockSize", "32k").toInt
-    new SnappyOutputStream(s, blockSize)
+    new SnappyOutputStreamWrapper(new SnappyOutputStream(s, blockSize))
   }
 
   override def compressedInputStream(s: InputStream): InputStream = new SnappyInputStream(s)
 }
+
+/**
+ * Wrapper over [[SnappyOutputStream]] which guards against write-after-close and double-close
+ * issues. See SPARK-7660 for more details. This wrapping can be removed if we upgrade to a version
+ * of snappy-java that contains the fix for https://github.com/xerial/snappy-java/issues/107.
+ */
+private final class SnappyOutputStreamWrapper(os: SnappyOutputStream) extends OutputStream {
+
+  private[this] var closed: Boolean = false
```

Comment:
I think we already do per-record checking of volatile vars.

https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/TaskContextImpl.scala#L44

On Fri, May 15, 2015 at 4:40 PM, Josh Rosen <notifications@github.com>
wrote:

> In core/src/main/scala/org/apache/spark/io/CompressionCodec.scala
> <https://github.com/apache/spark/pull/6176#discussion_r30454371>:
>
> >    }
> >
> >    override def compressedInputStream(s: InputStream): InputStream = new SnappyInputStream(s)
> >  }
> > +
> > +/**
> > + * Wrapper over [[SnappyOutputStream]] which guards against write-after-close and double-close
> > + * issues. See SPARK-7660 for more details. This wrapping can be removed if we upgrade to a version
> > + * of snappy-java that contains the fix for https://github.com/xerial/snappy-java/issues/107.
> > + */
> > +private final class SnappyOutputStreamWrapper(os: SnappyOutputStream) extends OutputStream {
> > +
> > +  private[this] var closed: Boolean = false
>
> Ah, good point. I guess this needs to be volatile in case we're performing
> cleanup in another thread. @rxin <https://github.com/rxin>, if this is
> volatile, won't that make the write() checks way more expensive?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/6176/files#r30454371>.
>